src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp

```diff
@@ -1733,6 +1755,83 @@ void CompositionalMultiphaseBase::chopNegativeDensities( DomainPartition & domai
+      arrayView1d< real64 const > const pres_n = subRegion.getExtrinsicData< extrinsicMeshData::flow::pressure_n >();
+      arrayView1d< real64 const > const temp = subRegion.getExtrinsicData< extrinsicMeshData::flow::temperature >();
+      arrayView1d< real64 const > const temp_n = subRegion.getExtrinsicData< extrinsicMeshData::flow::temperature_n >();
+      arrayView2d< real64 const, compflow::USD_PHASE > const phaseVolFrac =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::phaseVolumeFraction >();
+      arrayView2d< real64 const, compflow::USD_PHASE > const phaseVolFrac_n =
+        subRegion.getExtrinsicData< extrinsicMeshData::flow::phaseVolumeFraction_n >();
+
+      RAJA::ReduceMax< parallelDeviceReduce, real64 > subRegionMaxPresChange( 0.0 );
+      RAJA::ReduceMax< parallelDeviceReduce, real64 > subRegionMaxTempChange( 0.0 );
+      RAJA::ReduceMax< parallelDeviceReduce, real64 > subRegionMaxPhaseVolFracChange( 0.0 );
+
+      forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOSX_HOST_DEVICE ( localIndex const ei )
+      {
+        if( ghostRank[ei] < 0 )
+        {
+          subRegionMaxPresChange.max( LvArray::math::abs( pres[ei] - pres_n[ei] ) );
+          subRegionMaxTempChange.max( LvArray::math::abs( temp[ei] - temp_n[ei] ) );
+          for( integer ip = 0; ip < numPhase; ++ip )
+          {
+            subRegionMaxPhaseVolFracChange.max( LvArray::math::abs( phaseVolFrac[ei][ip] - phaseVolFrac_n[ei][ip] ) );
+          }
+        }
+      } );
+
+      maxAbsolutePressureChange = LvArray::math::max( maxAbsolutePressureChange, subRegionMaxPresChange.get() );
+      maxAbsoluteTemperatureChange = LvArray::math::max( maxAbsoluteTemperatureChange, subRegionMaxTempChange.get() );
+      maxAbsolutePhaseVolFracChange = LvArray::math::max( maxAbsolutePhaseVolFracChange, subRegionMaxPhaseVolFracChange.get() );
+
+    } );
+  } );
+
+  maxAbsolutePressureChange = MpiWrapper::max( maxAbsolutePressureChange );
+  maxAbsolutePhaseVolFracChange = MpiWrapper::max( maxAbsolutePhaseVolFracChange );
+  GEOSX_LOG_LEVEL_RANK_0( 1, getName() << ": Max absolute pressure change: "<< maxAbsolutePressureChange << " Pa" );
+  GEOSX_LOG_LEVEL_RANK_0( 1, getName() << ": Max absolute phase volume fraction change: "<< maxAbsolutePhaseVolFracChange );
+
+  if( m_isThermal )
+  {
+    maxAbsoluteTemperatureChange = MpiWrapper::max( maxAbsoluteTemperatureChange );
+    GEOSX_LOG_LEVEL_RANK_0( 1, getName() << ": Max absolute temperature change: "<< maxAbsoluteTemperatureChange << " K" );
+  }
+
+  real64 const eps = LvArray::NumericLimits< real64 >::epsilon;
+
+  real64 const nextDtPressure = currentDt *  ( 1.0 + m_solutionChangeScalingFactor ) * m_targetPressureChange
+                                / std::max( eps, maxAbsolutePressureChange + m_solutionChangeScalingFactor * m_targetPressureChange );
+  real64 const nextDtPhaseVolFrac = currentDt *  ( 1.0 + m_solutionChangeScalingFactor ) * m_targetPhaseVolFracChange
```

Comment:
Is this time of timestepping very effective for compositional flow? I have kind of a hard time guessing what good values are for max changes and for the `m_solutionChangeScalingFactor`. I am guessing it's mostly based on previous experience.

Reply:
Here I tried to follow the standard heuristics implemented in other reservoir simulators to avoid re-inventing the wheel. The implementation matches what is implemented in other reservoir simulators like MRST, AD-GPRS and the commercial ones (they all follow Hui Cao's thesis). 

The default values I picked are also inspired by their default values, with the most important ones being the `targetPhaseVolFractionChange` equal to 0.2 in most simulators. As for the `solutionChangeScalingFactor`, it has a minor impact and the default value of 0 is good enough, I think